src/prox_augmented_block.jl

```diff
@@ -0,0 +1,248 @@
+    vals::Union{AbstractVector,Nothing} = nothing,
+)
+
+    nlp = m.subproblem.problem_block
+    H =
+        sparse(
+            hess_structure(nlp)[1],
+            hess_structure(nlp)[2],
+            ones(nlp.meta.nnzh),
+            nlp.meta.nvar,
+            nlp.meta.nvar,
+        ) +
+        m.ρ .* abs.(m.A[:, m.subproblem.var_idx]' * m.A[:, m.subproblem.var_idx]) +
+        abs.(m.P)
+    rows, cols, temp_vals = findnz(H)
+    if isnothing(x)
+        return rows, cols
+    elseif isnothing(y)
+        H =
+            sparse(
+                hess_structure(nlp)[1],
+                hess_structure(nlp)[2],
+                hess_coord(nlp, x, obj_weight = obj_weight),
+                nlp.meta.nvar,
+                nlp.meta.nvar,
+            ) +
+            m.ρ .* m.A[:, m.subproblem.var_idx]' * m.A[:, m.subproblem.var_idx] +
+            m.P
+        for i = 1:length(rows)
+            vals[i] = H[rows[i], cols[i]]
+        end
+    else
+        H =
+            sparse(
+                hess_structure(nlp)[1],
+                hess_structure(nlp)[2],
+                hess_coord(nlp, x, y, obj_weight = obj_weight),
+                nlp.meta.nvar,
+                nlp.meta.nvar,
+            ) +
+            m.ρ .* m.A[:, m.subproblem.var_idx]' * m.A[:, m.subproblem.var_idx] +
+            m.P
+        for i = 1:length(rows)
+            vals[i] = H[rows[i], cols[i]]
+        end
+    end
+    return H
+end
+
+"""
+    update_primal!(
+      nlp::AugmentedNLPBlockModel, 
+      sol::AbstractVector
+    )
+Updates the primal solution estimate for the augmented nlp block `nlp`.
+# Arguments
+- `nlp::AugmentedNLPBlockModel`: the subproblem 
+- `sol::AbstractVector`: vector of primal variables
+"""
+function update_primal!(nlp::ProxAugmentedNLPBlockModel, sol::AbstractVector)
+    nlp.sol .= sol
+end
+
+"""
+    update_dual!(
+      nlp::AugmentedNLPBlockModel, 
+      λ::AbstractVector, 
+    )
+Updates the dual solution estimate in-place for the augmented nlp block `nlp`.
+# Arguments
+- `nlp::AugmentedNLPBlockModel`: the subproblem 
+- `λ::AbstractVector`: vector of dual variables
+"""
+function update_dual!(nlp::ProxAugmentedNLPBlockModel, λ::AbstractVector)
+    nlp.λ .= λ
+end
+
+function NLPModels.obj(nlp::ProxAugmentedNLPBlockModel, x::AbstractVector)
+    local_sol = deepcopy(nlp.sol)
+    local_sol[nlp.subproblem.var_idx] = x
+
+    return obj(nlp.subproblem.problem_block, x) +
+           dot(nlp.λ, nlp.A[:, nlp.subproblem.var_idx], x) +
+           (nlp.ρ / 2) * norm(nlp.A * local_sol - nlp.b)^2 +
+           1 / 2 * dot(
+               (x - nlp.sol[nlp.subproblem.var_idx]),
+               nlp.P,
+               (x - nlp.sol[nlp.subproblem.var_idx]),
+           )
+end
+
+function NLPModels.grad!(nlp::ProxAugmentedNLPBlockModel, x::AbstractVector, g::AbstractVector)
+    local_sol = deepcopy(nlp.sol)
+    local_sol[nlp.subproblem.var_idx] = x
+
+    grad!(nlp.subproblem.problem_block, x, g)
+    g .+=
```

Comment:
I would suggest using `mul!` (to dispatch on BLAS operations) instead of the broadcast operator here